### PR TITLE
8332920: C2: Partial Peeling is wrongly applied for CmpU with negative limit

### PR DESCRIPTION
**Pull request by chhagedorn**
A signed test is wrongly split off an unsigned test during Partial Peeling which results in not entering a loop even though it should. 

#### Idea of Partial Peeling

Partial Peeling rotates a loop with the hope to being able to convert the loop into a counted loop later. It is therefore preferable to use signed tests over unsigned tests because counted loops must use a signed loop exit test (i.e. with a `BaseCountedLoopEnd`). 

#### Partial Peeling with Unsigned Test

However, if there is only a suitable unsigned test, we can still try to use it for Partial Peeling. The idea is to then split off a signed version of the unsigned test and place it right before the unsigned test which can then be used as a loop exit test and later as `BaseCountedLoopEnd`:

https://github.com/openjdk/jdk/blob/0ea3bacf48be90d93f9e6c8e6568a0b8c61afb46/src/hotspot/share/opto/loopopts.cpp#L3074-L3080

#### Requirements for Using an Unsigned Test

The Signed and Unsigned Loop Exit Test do not need have the same result at runtime. It is sufficient if the Signed Loop Exit Test implies the Unsigned Loop Exit Test:
- If the Signed Loop Exit Test is false, then the (original) Unsigned Loop Exit Test will make sure to exit the loop if required.
- If the Signed Loop Exit Test is true, then the (original) Unsigned Loop Exit Test must also be true. Otherwise, we would exit a loop that we should have continued to execute. 

#### The Requirements Are Currently Broken

This strong requirement for splitting off a signed test is currently broken as seen in the test cases (for example, `testWhileLTIncr()`): We split off the signed loop exit test `i >= limit`, then partial peel and we get the signed loop exit test `i >= limit` as entry guard to the loop which is wrong:

https://github.com/openjdk/jdk/blob/0ea3bacf48be90d93f9e6c8e6568a0b8c61afb46/test/hotspot/jtreg/compiler/loopopts/TestPartialPeelAtUnsignedTestsNegativeLimit.java#L159-L178

#### Why Are the Requirements Broken?

The reason is that 
```
i >=u limit
```
can only be converted into the two signed comparisons
```
i < 0 || i >= limit
```
if `limit` is non-negative (i.e. `limit >= 0`): 
https://github.com/openjdk/jdk/blob/0ea3bacf48be90d93f9e6c8e6568a0b8c61afb46/src/hotspot/share/opto/loopopts.cpp#L3054-L3061

This is currently missing and we wrongly use `i < 0` or `i >= limit` as split off signed test.

#### Fixing the Broken Requirements
To fix this, I've added a bailout when `limit` could be negative which is the same as checking if the type of `limit` contains a negative value:

https://github.com/openjdk/jdk/blob/0ea3bacf48be90d93f9e6c8e6568a0b8c61afb46/src/hotspot/share/opto/loopopts.cpp#L3062-L3066

Thanks,
Christian

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332920](https://bugs.openjdk.org/browse/JDK-8332920): C2: Partial Peeling is wrongly applied for CmpU with negative limit (**Bug** - P2)(⚠️ The fixVersion in this issue is [23] but the fixVersion in .jcheck/conf is 24, a new backport will be created when this pr is integrated.)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [e0a18f06](https://git.openjdk.org/jdk/pull/19522/files/e0a18f0625ae9139b624a7f50ff21f8116f07cb9)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [33d89249](https://git.openjdk.org/jdk/pull/19522/files/33d89249886f434d8c96104f4b934e72c31d9d52)
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19522/head:pull/19522` \
`$ git checkout pull/19522`

Update a local copy of the PR: \
`$ git checkout pull/19522` \
`$ git pull https://git.openjdk.org/jdk.git pull/19522/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19522`

View PR using the GUI difftool: \
`$ git pr show -t 19522`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19522.diff">https://git.openjdk.org/jdk/pull/19522.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19522#issuecomment-2145162981)

### CODE REVIEW

**Comment by chhagedorn** [comment]
Thanks Vladimir for your review!

**Review by TobiHartmann — COMMENTED** [review]
> If the Signed Loop Exit Test is false, then the (original) Unsigned Loop Exit Test will make sure to exit the loop if required.

Is that comment incorrect then?

https://github.com/openjdk/jdk/blob/2edb6d98133d8bd6dc4527c7497c460283fdc53e/src/hotspot/share/opto/loopopts.cpp#L2993-L2994

File: src/hotspot/share/opto/loopopts.cpp [hunk 1]
```diff
@@ -3043,22 +3043,80 @@ IfNode* PhaseIdealLoop::insert_cmpi_loop_exit(IfNode* if_cmpu, IdealLoopTree *lo
...
-  // The loop exit condition is !(i <u limit) ==> (i < 0 || i >= limit).
-  // Split out the exit condition (i < 0) for stride < 0 or (i >= limit) for stride > 0.
-  Node* limit = nullptr;
+  // The unsigned loop exit condition is
+  //   !(i <u  limit)
+  // =   i >=u limit
+  // If
+  //   limit >= 0   (COND)
+  // then the unsigned loop exit condition is equivalent to the signed loop exit condition
+  //   i < 0 || i >= limit
```
Comment (eme64) [inline src/hotspot/share/opto/loopopts.cpp:3052]: Could we come up with an alternative equation if `limit < 0`? And maybe even a combined condition? Not sure if that is helpful, but I'd like to think about it. Could also be a follow-up RFE.

File: src/hotspot/share/opto/loopopts.cpp [hunk 2]
```diff
@@ -3043,22 +3043,80 @@ IfNode* PhaseIdealLoop::insert_cmpi_loop_exit(IfNode* if_cmpu, IdealLoopTree *lo
...
+  //   Unsigned Loop Exit Condition i >=u limit
+  //   <rest of unpeeled section>
+  //   goto Loop
+  //
+  // We either exit the loop with the Unsigned Loop Exit Condition if the Signed Loop Exit Condition is false, or we
+  // exit the loop with the Signed Loop Exit Condition. In the latter case, the Unsigned Loop Exit Condition also needs
+  // to be true (otherwise, we wrongly exit a loop that should not have been exited). More formally, we need to ensure:
+  //   "Signed Loop Exit Test" implies "Unsigned Loop Exit Test"
+  // This is trivially given:
+  // - Stride < 0:
```
Comment (eme64) [inline src/hotspot/share/opto/loopopts.cpp:3087]:
```suggestion
  // - stride < 0:
```
Stylistic decision, leave this to your

File: src/hotspot/share/opto/loopopts.cpp [hunk 3]
```diff
@@ -3043,22 +3043,80 @@ IfNode* PhaseIdealLoop::insert_cmpi_loop_exit(IfNode* if_cmpu, IdealLoopTree *lo
...
+  //   i  >=u limit
+  //   -3 >=u -2
+  // is false.
+  Node* limit = cmpu->in(2);
+  const TypeInt* type_limit = _igvn.type(limit)->is_int();
+  if (type_limit->_lo < 0) {
+    return nullptr;
+  }
+
+  // For stride < 0, we split off the signed loop exit condition
```
Comment (eme64) [inline src/hotspot/share/opto/loopopts.cpp:3068]:
```suggestion
  // For stride < 0, we insert off the signed loop exit condition
```
Why do you say "split", we are inserting this extra check, right? Or is the idea that the "rotation" puts this new condition as the last check, hence the exit check in the loop body? Being a big more explicit could help here.

File: src/hotspot/share/opto/loopopts.cpp [hunk 4]
```diff
@@ -3043,22 +3043,80 @@ IfNode* PhaseIdealLoop::insert_cmpi_loop_exit(IfNode* if_cmpu, IdealLoopTree *lo
...
+  //    which is the Unsigned Loop Exit Condition.
+  //
+  // After Partial Peeling, we have the following structure:
+  //   <cloned peeled section>
+  //   Signed Loop Exit Condition i < 0 or i >= limit
+  //   Loop:
+  //     Unsigned Loop Exit Condition i >=u limit
+  //     <rest of unpeeled section>
+  //     <peeled section>
+  //     Signed Loop Exit Condition i < 0 or i >= limit
```
Comment (eme64) [inline src/hotspot/share/opto/loopopts.cpp:3104]: There is not litterally an `OR` here, right? It's a bit confusing.

**Review by eme64 — COMMENTED** [review]
More to come later

File: src/hotspot/share/opto/loopopts.cpp [hunk 5]
```diff
@@ -3043,22 +3043,80 @@ IfNode* PhaseIdealLoop::insert_cmpi_loop_exit(IfNode* if_cmpu, IdealLoopTree *lo
...
+  //      i >=u limit    // limit <= MAX_INT (trivially) and since limit >= 0 assumption (COND)
+  //    which is the Unsigned Loop Exit Condition.
+  // - Stride > 0:
+  //      i >=  limit   // Signed Loop Exit Condition
+  //      i >=u limit   // Since limit >= 0 assumption (COND)
+  //    which is the Unsigned Loop Exit Condition.
+  //
+  // After Partial Peeling, we have the following structure:
+  //   <cloned peeled section>
+  //   Signed Loop Exit Condition i < 0 or i >= limit
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopopts.cpp:3099]:
```suggestion
  //   Signed Loop Exit Condition i < 0 (or i >= limit)
```

File: src/hotspot/share/opto/loopopts.cpp (hunk 4 above)
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopopts.cpp:3104]:
```suggestion
  //     Signed Loop Exit Condition i < 0 (or i >= limit)
```

File: src/hotspot/share/opto/loopopts.cpp [hunk 6]
```diff
@@ -3043,22 +3043,80 @@ IfNode* PhaseIdealLoop::insert_cmpi_loop_exit(IfNode* if_cmpu, IdealLoopTree *lo
...
+
+  // For stride < 0, we split off the signed loop exit condition
+  //   i < 0
+  // and for stride > 0
+  //   i >= limit
+  // such that we have the following graph before Partial Peeling:
+  //
+  // Loop:
+  //   <peeled section>
+  //   Signed Loop Exit Condition i < 0 or i >= limit
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopopts.cpp:3076]:
```suggestion
  //   Signed Loop Exit Condition i < 0 (or i >= limit)
```

File: src/hotspot/share/opto/loopopts.cpp [hunk 7]
```diff
@@ -3043,22 +3043,80 @@ IfNode* PhaseIdealLoop::insert_cmpi_loop_exit(IfNode* if_cmpu, IdealLoopTree *lo
...
-  Node* limit = nullptr;
+  // The unsigned loop exit condition is
+  //   !(i <u  limit)
+  // =   i >=u limit
+  // If
+  //   limit >= 0   (COND)
+  // then the unsigned loop exit condition is equivalent to the signed loop exit condition
+  //   i < 0 || i >= limit
+  //
+  // Note that this does not hold for limit < 0:
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopopts.cpp:3054]: As we discussed, rephrase this as counterexample.

File: src/hotspot/share/opto/loopopts.cpp [hunk 8]
```diff
@@ -3043,22 +3043,80 @@ IfNode* PhaseIdealLoop::insert_cmpi_loop_exit(IfNode* if_cmpu, IdealLoopTree *lo
...
+  //
+  // We either exit the loop with the Unsigned Loop Exit Condition if the Signed Loop Exit Condition is false, or we
+  // exit the loop with the Signed Loop Exit Condition. In the latter case, the Unsigned Loop Exit Condition also needs
+  // to be true (otherwise, we wrongly exit a loop that should not have been exited). More formally, we need to ensure:
+  //   "Signed Loop Exit Test" implies "Unsigned Loop Exit Test"
+  // This is trivially given:
+  // - Stride < 0:
+  //      i <   0        // Signed Loop Exit Condition
+  //      i >u  MAX_INT  // all negative values are greater than MAX_INT when converted to unsigned
+  //      i >=u limit    // limit <= MAX_INT (trivially) and since limit >= 0 assumption (COND)
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopopts.cpp:3090]: As we discussed, an intermediate step in the proof would be good here.

**Review by TobiHartmann — APPROVED** [review]
Looks great otherwise. Nice test and proof!

**Review by TobiHartmann — APPROVED** [review]
Nice! That looks good to me.

**Comment by chhagedorn** [comment]
Thanks Tobias for your careful review!

File: src/hotspot/share/opto/loopopts.cpp [hunk 9]
```diff
@@ -2986,52 +2986,98 @@ RegionNode* PhaseIdealLoop::insert_region_before_proj(ProjNode* proj) {
...
+//       /  /       |
+//      /  /        v
+//     /  /  if (i <u limit)          Cloned unsigned loop exit test
+//    /  /   /      |
+//   v  v  v        |
+//  exit-region     |
+//        |         |
+//    dummy-if      |
+//     /  |         |
+// other  |         |
```
Comment (eme64) [inline src/hotspot/share/opto/loopopts.cpp:3057]: what is this?

File: src/hotspot/share/opto/loopopts.cpp [hunk 10]
```diff
@@ -3043,22 +3089,101 @@ IfNode* PhaseIdealLoop::insert_cmpi_loop_exit(IfNode* if_cmpu, IdealLoopTree *lo
...
+  //   i  >=u limit
+  //   -3 >=u -2
+  // is false.
+  Node* limit = cmpu->in(2);
+  const TypeInt* type_limit = _igvn.type(limit)->is_int();
+  if (type_limit->_lo < 0) {
+    return nullptr;
+  }
+
+  // From (SLE-full), we can extract a single signed loop exit condition depending on the stride:
```
Comment (eme64) [inline src/hotspot/share/opto/loopopts.cpp:3128]:
```suggestion
  // We prove below that we can extract a single signed loop exit condition from (SLE-full), depending on the stride:
```

File: src/hotspot/share/opto/loopopts.cpp [hunk 11]
```diff
@@ -3043,22 +3089,101 @@ IfNode* PhaseIdealLoop::insert_cmpi_loop_exit(IfNode* if_cmpu, IdealLoopTree *lo
...
+  //   stride < 0:
+  //     i < 0        (SLE-negative)
+  //   stride > 0:
+  //     i >= limit   (SLE-positive)
```
Comment (eme64) [inline src/hotspot/share/opto/loopopts.cpp:3132]:
```suggestion
  //   stride < 0:
  //     i < 0        (SLE = SLE-negative)
  //   stride > 0:
  //     i >= limit   (SLE = SLE-positive)
```
That would give you a definition of (SLE).

File: src/hotspot/share/opto/loopopts.cpp [hunk 12]
```diff
@@ -3043,22 +3089,101 @@ IfNode* PhaseIdealLoop::insert_cmpi_loop_exit(IfNode* if_cmpu, IdealLoopTree *lo
...
+  // exit a loop that should not have been exited if we did not apply Partial Peeling. More formally, we need to ensure:
+  //   (SLE) IMPLIES (ULE)
+  // This indeed holds when (COND) is given:
+  // - stride > 0:
+  //       i >=  limit             // (SLE-positive)
+  //       i >=  limit >= 0        // (COND)
+  //       i >=u limit >= 0        // (LEMMA)
+  //     which is the unsigned loop exit condition (ULE).
+  // - stride < 0:
+  //       i        <  0           // (SLE-negative)
```
Comment (eme64) [inline src/hotspot/share/opto/loopopts.cpp:3155]:
```suggestion
  //       i >=  limit             // (SLE = SLE-positive)
  //       i >=  limit >= 0        // (COND)
  //       i >=u limit >= 0        // (LEMMA)
  //     which is the unsigned loop exit condition (ULE).
  // - stride < 0:
  //       i        <  0           // (SLE = SLE-negative)
```

File: src/hotspot/share/opto/loopopts.cpp [hunk 13]
```diff
@@ -3043,22 +3089,101 @@ IfNode* PhaseIdealLoop::insert_cmpi_loop_exit(IfNode* if_cmpu, IdealLoopTree *lo
...
+  //
+  //
+  // After Partial Peeling, we have the following structure for stride > 0 (similar for stride < 0):
+  //   <cloned peeled section>
+  //   i >= limit (SLE)
+  //   Loop:
+  //     i >=u limit (ULE)
+  //     <rest of unpeeled section>
+  //     <peeled section>
+  //     i >= limit (SLE)
```
Comment (eme64) [inline src/hotspot/share/opto/loopopts.cpp:3171]:
```suggestion
  //   i >= limit (SLE-positive)
  //   Loop:
  //     i >=u limit (ULE)
  //     <rest of unpeeled section>
  //     <peeled section>
  //     i >= limit (SLE-positive)
```

File: test/hotspot/jtreg/compiler/loopopts/TestPartialPeelAtUnsignedTestsNegativeLimit.java [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+        testWhileLTDecr(MIN_VALUE, -2000);
+        check(MIN_VALUE + 1); // MAX_VALUE + 2 iterations
+        testWhileLTDecr(MIN_VALUE, MIN_VALUE + 1);
+        check(MIN_VALUE + 1); // MAX_VALUE + 2 iterations
+        testWhileLTDecr(MIN_VALUE + 2000, -1);
+        check(MIN_VALUE + 2001); // MAX_VALUE + 2002 iterations
+        testWhileLTDecr(MIN_VALUE + 2000, -2000);
+        check(MIN_VALUE + 2001); // MAX_VALUE + 2002 iterations
+        testWhileLTDecr(MIN_VALUE + 2000, MIN_VALUE + 2001);
+        check(MIN_VALUE + 2001); // MAX_VALUE + 2002 iterations
```
Comment (eme64) [inline test/hotspot/jtreg/compiler/loopopts/TestPartialPeelAtUnsignedTestsNegativeLimit.java:270]: Could you also add a randomized input test here, that plays close to the boundaries? Just to make sure we would catch things like some off-by one errors.

Comment (chhagedorn) [reply test/hotspot/jtreg/compiler/loopopts/TestPartialPeelAtUnsignedTestsNegativeLimit.java:287]: Good idea, I've added some random test and computed the number of iteration that I expect. I only did it for the interesting cases.

**Review by vnkozlov — APPROVED** [review]
Update is good.

**Review by eme64 — APPROVED** [review]
Nice work with the proofs, good work @chhagedorn !

File: src/hotspot/share/opto/loopopts.cpp [hunk 15]
```diff
@@ -2986,52 +2986,98 @@ RegionNode* PhaseIdealLoop::insert_region_before_proj(ProjNode* proj) {
...
+//         /        v
+//        /  if (i <  limit)          Split off signed loop exit test
+//       /  /       |
+//      /  /        v
+//     /  /  if (i <u limit)          Cloned unsigned loop exit test
+//    /  /   /      |
+//   v  v  v        |
+//  exit-region     |
+//        |         |
+//    dummy-if      |
```
[the remark this answers is not in the document]
Comment (chhagedorn) [reply src/hotspot/share/opto/loopopts.cpp:3055]: Fair point, added a note.

**Comment by chhagedorn** [comment]
Thanks Emanuel and also for the review!

**Comment by chhagedorn** [comment]
Performance testing looked good. Thanks again for your careful reviews!

/integrate
